zerver/lib/users.py

```diff
@@ -526,6 +542,51 @@ def check_user_can_access_all_users(acting_user: Optional[UserProfile]) -> bool:
     return False
 
 
+def check_can_access_user(
+    target_user: UserProfile, user_profile: Optional[UserProfile] = None
+) -> bool:
+    if not user_access_restricted_in_realm(target_user):
+        return True
+
+    if check_user_can_access_all_users(user_profile):
+        return True
+
+    assert user_profile is not None
+
+    if target_user.id == user_profile.id:
+        return True
+
+    # These include Subscription objects for streams as well as group DMs.
+    subscribed_recipient_ids = Subscription.objects.filter(
+        user_profile=user_profile,
+        active=True,
+        recipient__type__in=[Recipient.STREAM, Recipient.HUDDLE],
+    ).values_list("recipient_id", flat=True)
+
+    if Subscription.objects.filter(
+        recipient_id__in=subscribed_recipient_ids,
+        user_profile=target_user,
+        active=True,
+        is_user_active=True,
```

Comment:
One thing as mentioned in the commit message is that we would not allow guests to mute or unmute deactivated users who were part of the streams that the guest is subscribed to but were not involved in DMs with them.

Reply:
The muting aspect is weird but likely not actually important. I'll add a comment to keep track of this observation.